nipype/pipeline/engine/workflows.py

```diff
@@ -314,6 +320,8 @@ def disconnect(self, *args):
             else:
                 self._graph.add_edges_from([(srcnode, dstnode, edge_data)])
 
+        self._update_node_cache()
+
```

Comment:
Looking at your latest profile, you spend a chunk of time in `disconnect()`, but it does not remove nodes, only edges.

```suggestion
```